src/duk_debugger.c

```diff
@@ -1448,6 +1448,49 @@ DUK_LOCAL void duk__debug_handle_detach(duk_hthread *thr, duk_heap *heap) {
 	DUK__SET_CONN_BROKEN(thr, 0);  /* not an error */
 }
 
+DUK_LOCAL void duk__debug_handle_apprequest(duk_hthread *thr, duk_heap *heap) {
+	DUK_D(DUK_DPRINT("debug command AppRequest"));
+
+	if (heap->dbg_request_cb != NULL) {
+		duk_context *ctx = (duk_context *) thr;
+		duk_idx_t nrets;
+		duk_idx_t nvalues = 0;
+		duk_idx_t old_top;
+		duk_idx_t top, idx;
+
+		/* Read tvals from the message and push them onto the valstack,
+		 * then call the request callback to process the request.
+		 */
+		old_top = duk_get_top(ctx);  /* save stack top */
+		while (duk_debug_peek_byte(thr) != DUK_DBG_MARKER_EOM) {
+			duk_debug_read_tval(thr);  /* push to stack */
+			nvalues++;
+		}
+
+		/* Request callback should push values for reply to client onto valstack */
+		nrets = heap->dbg_request_cb(heap->dbg_udata, ctx, nvalues);
+		if (nrets > 0) {
+			DUK_ASSERT(duk_get_top(ctx) >= old_top + nrets);
+
+			/* Reply with tvals pushed by request callback */
+			duk_debug_write_byte(thr, DUK_DBG_MARKER_REPLY);
+			top = duk_get_top(ctx);
+			for (idx = top - nrets; idx < top; idx++) {
+				duk_debug_write_tval(thr, DUK_GET_TVAL_POSIDX(ctx, idx));
```

Comment:
Hmm. Because the `nrets` check is just an assert this should probably be `duk_get_tval(ctx, idx)`. However, that doesn't really improve the situation very much because `duk_debug_write_tval()` will assert for `tv != NULL`.

Maybe the `duk_get_top(ctx) >= old_top + nrets` should be an actual if-check:

    if (duk_get_top(ctx) < old_top + nrets) {
        DUK_D(DUK_DPRINT("AppRequest callback return value doesn't match value stack configuration"));
        goto fail;
    }

The fail label could be shared with the duk_check_stack() and simply do a hard detach.

Anyway, with this change `DUK_GET_TVAL_POSIDX()` can remain and "buys space" for the nrets check.